Batched local store

domsj opened the pull request.

domsj committed:
wrapping Local_store in Batched_store

domsj committed:
implemented batched store

domsj committed:
small cleanup of STORE module type

domsj committed:
verifying current transaction in batched_store and mem_store

domsj committed:
added 008_install_dev_env.sh to jenkins/{quick/left/right}

domsj committed:
fixed delete_prefix and rev_range in batched_store

domsj committed:
fixed catchup only test

domsj committed:
fixed potential bug which could cause an individual update to be applied twice

NicolasT commented:
TODO:
- [x] Count key/value-sizes inserted in batch, and commit batch when some configurable threshold is reached.
- [x] Make entry count for batch configurable.

toolslive commented on src/node/node_main.ml:
```diff
@@ -199,7 +199,8 @@ let only_catchup (type s) (module S : Store.STORE with type t = s) ~name ~cluste
   Catchup.catchup me other_configs ~cluster_id 
     ((module S),store,tlc)  current_i mr_name (future_n,future_i) >>= fun _ ->
-  tlc # close () 
+  tlc # close ();
```
tlc # close  () >>= ...

NicolasT replied:
Might want to enable warnings-as-errors during build?

domsj replied:
after discussion in the chat you're on top of this in a soon to be other pull request :)

toolslive commented on src/node/node_main.ml:
```diff
@@ -651,7 +652,7 @@ let _main_2 (type s)
       
 let main_t make_config name daemonize catchup_only : int Lwt.t =
-  let module S = (val (Store.make_store_module (module Local_store))) in
+  let module S = (val (Store.make_store_module (module Batched_store.Local_store))) in
```
maybe make this configurable....

NicolasT replied:
If the number of 'batched items' is configurable, setting this to 1 would yield the old behavior of using plain Local_store, without the need to be able to disable Batched_store. Would this be suitable?

toolslive commented on src/node/batched_store.ml:
```diff
@@ -0,0 +1,279 @@
+  let copy_store2 =
+    S.copy_store2
+
+  let get_fringe s =
```
get_fringe, copy_store, defrag, optimize, get_key_count, get_location,....
all look very similar. maybe it's worth to extract the commonalities.

domsj replied:
I have extracted (most of) the commonalities already into _sync_and_start_transaction_if_needed ;-).
I don't think any more commonalities are worth the trouble so I'll leave this as is.

toolslive commented on src/node/batched_store.ml:
```diff
@@ -0,0 +1,279 @@
+      | Some ls_tx ->
+          _apply_vo_to_local_store s.s ls_tx k vo
+
+  let _finalize f g =
```
unused. (and pretty dangerous )

NicolasT replied:
:-1: This doesn't look OK, better get rid of it.

domsj replied:
Agreed, removing

NicolasT commented on src/node/batched_store.ml:
```diff
@@ -0,0 +1,279 @@
+          if tx != tx'
+          then failwith "the provided transaction is not the current transaction of the batched store"
+
+  let exists s k =
```
'exists' and 'get' share quite some logic. Consider lifting this?

NicolasT commented on src/node/store.ml:
```diff
@@ -547,7 +537,12 @@ struct
       | Key key ->
-          with_transaction store ~key:(Some key) f
+          let matched_locks = match store._tx_lock with
+            | Some txl -> txl == key
```
'==' vs '='?

domsj replied:
== tests for physical equality, which is what I want here. I'll add a comment to the code to make this more explicit.

NicolasT commented on src/node/store_test.ml:
```diff
@@ -125,23 +135,11 @@ let test_safe_insert_value_with_partial_value_update () =
     Lwt.return ()
   )
 
-let test_safe_insert_value_with_tx () =
```
Care to elaborate why this test was removed?

domsj replied:
This test verified the behaviour of S.safe_insert_value with a ~tx:(Some a_tx) passed to it. As you can see at https://github.com/Incubaid/arakoon/pull/61/files#L11R818 this possibility has been removed (it was there to allow batching from in catchup.ml).

domsj committed:
fixed some remarks from the pull request

domsj committed:
Batched_store max_entries and max_size is now configurable
Output of entire cluster_cfg on node startup
Fix for regression in batched_store of previous commit

NicolasT commented:
Is this ready for merge?
